File: backend/python/RobotImpl.py
#!/usr/bin/python

from IRobot import IRobot
from Maze import Maze
from Point import Point
from Reset import ResetException
import sys

class RobotImpl(object):
    start = Point(0,0)
    location = Point(1,1)
    target = Point(2,2)
    heading = IRobot.EAST

    def __init__(self, maze, prefix):
        self.prefix = prefix
        self.steps = 0
        self.collisions = 0
        self.setMaze(maze)

    def setMaze(self,maze):
        self.maze = maze
        self.width = maze.getWidth()
        self.height = maze.getHeight()
        self.start = self.maze.start
        self.target = self.maze.target
        self.location = self.maze.location
        self.maze.setTile(IRobot.BEENBEFORE, self.maze.location)
        self.heading = self.maze.heading
        self.runs = 0

    # Getter and Setter methods

    def getLocationX(self):
        return self.location.x
    def getLocationY(self):
        return self.location.y
    def getLocation(self):
        return self.location

    def getTargetLocation(self):
        return self.target
   
    def setTargetLocation(self, px, py=None):
        if(py!=None):
            self.target = Point(px,py)
        else:
            self.target = px

    def getHeading(self):
        return self.heading
    def setHeading(self, newHeading):
        if(newHeading < IRobot.NORTH or newHeading > IRobot.WEST):
            raise RuntimeError("The robot's heading can only be NORTH, SOUTH, EAST or WEST.")
        self.heading = newHeading

    def getSteps(self):
        return self.steps
    def getCollisions(self):
        return self.collisions
    def getRuns(self):
        return self.runs

    # Turn the robot to face a new direction
    def face(self, newdir):
        if(newdir < IRobot.AHEAD or newdir > IRobot.LEFT):
            raise RuntimeError("The robot can only face AHEAD, BEHIND, LEFT and RIGHT.")
        if(newdir == IRobot.BEHIND):
            self.setHeading((self.heading + 2) % 4 + IRobot.NORTH)
        elif(newdir ==  IRobot.LEFT):
            self.setHeading((self.heading + 3) % 4 + IRobot.NORTH)
        elif(newdir ==  IRobot.RIGHT):
            self.setHeading((self.heading + 1) % 4 + IRobot.NORTH)

    # Inspect the adjecent tile in a given direction to find what type it is
    def look(self, direction):
        if(direction < IRobot.AHEAD or direction > IRobot.LEFT):
            raise RuntimeError("The robot can only look AHEAD, BEHIND, LEFT and RIGHT.")
        
        newHeading = (self.heading + direction) % 4
        if(newHeading == 0):
            point = Point(self.getLocationX(), self.getLocationY() - 1)
        elif(newHeading == 1):
            point = Point(self.getLocationX() + 1, self.getLocationY())
        elif(newHeading == 2):
            point = Point(self.getLocationX(), self.getLocationY() + 1)
        elif(newHeading == 3):
            point = Point(self.getLocationX() - 1, self.getLocationY())
        else:
            point = Point(self.getLocationX(), self.getLocationY())

        tileType = self.maze.getTileType(point)

        return tileType

    # Attempt to move the robot forwards
    def advance(self):
        x = self.location.x
        y = self.location.y

        if(self.heading == IRobot.NORTH):
            y = y-1
        elif(self.heading == IRobot.EAST):
            x = x+1
        elif(self.heading == IRobot.SOUTH):
            y = y+1
        elif(self.heading == IRobot.WEST):
            x = x-1

        if (x<0 or y<0 or x>=self.maze.getWidth() or y>=self.maze.getHeight()):
            raise RuntimeError("Robot cannot advance off the edge of the maze!")

        if(self.maze.getTileType(x,y) == IRobot.WALL):
            self.collisions = self.collisions+1
        else:
            self.steps = self.steps+1
            self.location = Point(x,y)

        self.maze.setTile(IRobot.BEENBEFORE,self.location)

    # Reset the robot and the maze to their original state
    def reset(self):
        for x in range(self.width):
            for y in range(self.height):
                if self.maze.getTileType(x,y) == IRobot.BEENBEFORE:
                    self.maze.setTile(IRobot.PASSAGE,x,y)
        self.location = self.start
        self.heading = IRobot.EAST
        self.maze.setTile(IRobot.BEENBEFORE,self.location)
        self.steps = 0
        self.collisions = 0
        self.runs += 1

    # Output JSON representing the current state of the robot and maze to stdout
    def jsondump(self):
        # Wait for instruction from the middleware
        line=""
        while (line != "step\n" and line != "reset\n"):
            line = sys.stdin.readline()
        
        data = ''

        # If a reset instruction was received, reset the robot and maze and add a reset prefix to the return string
        if line == "reset\n":
            self.reset()
            data += '[RESET]'

        # maze is transposed when input, so transpose back for outputting
        m = [[r[col] for r in self.maze.maze] for col in range(len(self.maze.maze[0]))]
        # Construct a string containing JSON representing the current state of the robot and maze
        data += '{ "layout":'+str(m)+','
        data += '"start_pos":{"x":'+str(self.start.x)+',"y":'+str(self.start.y)+'},'
        data += '"finish_pos":{"x":'+str(self.target.x)+',"y":'+str(self.target.y)+'},'
        data += '"robot_pos":{"x":'+str(self.location.x)+',"y":'+str(self.location.y)+'},'
        data += '"robot_orientation":'+str(self.heading)+','
        data += '"steps":'+str(self.steps)+','
        data += '"collisions":'+str(self.collisions)+','
        data += '"goal_reached":'+str(self.location == self.getTargetLocation()).lower()+','
        data += '"runs":'+str(self.runs)+'}'

        # Print the string prepended with the prefix
        print(self.prefix+data)

        # If a reset instruction was received, adjust the number of runs to avoid increasing twice
        # then throw a ResetException to halt execution of the current run
        if line == "reset\n":
            self.runs -= 1
            raise ResetException("RESET")

File: backend/java/src/RobotImpl.java
import java.awt.Point;
import java.io.*;

public class RobotImpl implements IRobot {
    Maze maze;
    int height;
    int width;
    Point start;
    Point target;
    Point location;
    int heading;
    int steps;
    int collisions;
    int runs;
    String prefix;

    public RobotImpl(Maze m, String prefix) {
        this.prefix = prefix;
        setMaze(m);
        steps = 0;
        collisions = 0;
    }

    public void setMaze(Maze m) {
        maze = m;
        height = m.getHeight();
        width = m.getWidth();
        start = m.getStart();
        target = m.getFinish();
        location = m.getLocation();
        maze.setTile(IRobot.BEENBEFORE,start);
        heading = m.getHeading();
        runs = 0;
    }

    // Getter and Setter methods

    public int getLocationX() {
        return location.x;
    }
    public int getLocationY() {
        return location.y;
    }
    public Point getLocation() {
        return location;
    }

    public Point getTargetLocation() {
        return target;
    }
   
    public void setTargetLocation(int x, int y) {
        target = new Point(x,y);
    }
    public void setTargetLocation(Point p) {
        target = p;
    }

    public int getSteps() {
        return steps;
    }
    public int getCollisions() {
        return collisions;
    }
    public int getRuns() {
        return runs;
    }

    public int getHeading() {
        return heading;
    }
    public void setHeading(int newHeading) {
        if(newHeading < IRobot.NORTH || newHeading > IRobot.WEST) {
            throw new RuntimeException("The robot's heading can only be NORTH, SOUTH, EAST or WEST.");
        }
        heading = newHeading;
    }

    // Inspect the adjecent tile in a given direction to find what type it is
    public int look(int dir) throws RuntimeException {
        if (dir < IRobot.AHEAD || dir > IRobot.LEFT) {
            throw new RuntimeException("The robot can only look AHEAD, BEHIND, LEFT and RIGHT.");
        }
      
        int newHeading = heading + dir;
        Point point = new Point();
        switch (newHeading % 4) {
            case 0:
                point = new Point(getLocationX(), getLocationY() - 1);
                break;
            case 1:
                point = new Point(getLocationX() + 1, getLocationY());
                break;
            case 2:
                point = new Point(getLocationX(), getLocationY() + 1);
                break;
            case 3:
                point = new Point(getLocationX() - 1, getLocationY());
                break;
        }
        int type = maze.getTileType(point);
        return type;
    }

    // Turn the robot to face a new direction
    public void face(int dir) throws RuntimeException {
        if (dir < IRobot.AHEAD || dir > IRobot.LEFT) {
            throw new RuntimeException("The robot can only face AHEAD, BEHIND, LEFT and RIGHT.");
        }
        switch (dir) {
            case IRobot.BEHIND:
                setHeading((heading + 2) % 4 + IRobot.NORTH);
                break;
            case IRobot.LEFT:
                setHeading((heading + 3) % 4 + IRobot.NORTH);
                break;
            case IRobot.RIGHT:
                setHeading((heading + 1) % 4 + IRobot.NORTH);
                break;
        }
    }
    
    // Attempt to move the robot forwards
    public void advance() throws RuntimeException {
        int x = location.x;
        int y = location.y;
        switch (getHeading()) {
            case IRobot.NORTH:
                y--;
                break;
            case IRobot.EAST:
                x++;
                break;
            case IRobot.SOUTH:
                y++;
                break;
            case IRobot.WEST:
                x--;
                break;
        }
         
        if (x<0 || y<0 || x>=maze.getWidth() || y>=maze.getHeight()) {
            throw new RuntimeException("Robot cannot advance off the edge of the maze!");
        }
         
        if (maze.getTileType(x, y) == IRobot.WALL) {
            collisions++;
        } else {
            steps++;
            location = new Point(x,y);
        }
        maze.setTile(IRobot.BEENBEFORE,location);
    }

    // Reset the robot and the maze to their original state
    public void reset() {
        for (int x = 0; x < width; x++) {
            for (int y = 0; y < height; y++) {
                if (maze.getTileType(x,y) == IRobot.BEENBEFORE) {
                    maze.setTile(IRobot.PASSAGE,x,y);
                }
            }
        }
        location = start;
        heading = IRobot.EAST;
        maze.setTile(IRobot.BEENBEFORE,location);
        steps = 0;
        collisions = 0;
        runs++;
    }

    // Output JSON representing the current state of the robot and maze to stdout
    public void jsondump() throws ResetException {
        // Wait for instruction from the middleware
        String line = "";
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        while(!(line.equals("step") || line.equals("reset"))) {
            try {
                line = br.readLine();
            } catch (IOException ioe) {
                System.out.println(ioe.getMessage());
            }
        }
        
        String data = "";

        // If a reset instruction was received, reset the robot and maze and add a reset prefix to the return string
        if(line.equals("reset")) {
            data += "[RESET]";
            reset();
        }

        // maze is transposed when input, so transpose back for outputting
        int[][] m = maze.transpose(maze.maze);
        // Construct a string containing JSON representing the current state of the robot and maze
        String s = "";
        String mazeString = "[";
        for(int y = 0; y < maze.maze[0].length; y++) {
            mazeString += "[";
            for(int x = 0; x < maze.maze.length; x++) {
                mazeString += maze.maze[x][y];
                if (x == maze.maze.length - 1) {
                    s = "";
                } else {
                    s = ",";
                }
                mazeString += s;
            }
            if (y == maze.maze[0].length - 1) {
                s = "]";
            } else {
                s = ",";
            }
            mazeString += "]" + s;
        }
        data += "{ \"layout\":" + mazeString + ",";
        data += "\"start_pos\":{\"x\":" + start.x + ",\"y\":" + start.y + "},";
        data += "\"finish_pos\":{\"x\":" + target.x + ",\"y\":" + target.y + "},";
        data += "\"robot_pos\":{\"x\":" + location.x + ",\"y\":" + location.y + "},";
        data += "\"robot_orientation\":" + heading + ",";
        data += "\"steps\":" + steps + ",";
        data += "\"collisions\":" + collisions + ",";
        data += "\"goal_reached\":" + (location == getTargetLocation()) + ",";
        data += "\"runs\":" + runs + "}";
        // Print the string prepended with the prefix
        System.out.println(prefix + data);

        // If a reset instruction was received, adjust the number of runs to avoid increasing twice
        // then throw a ResetException to halt execution of the current run
        if(line.equals("reset")) {
            runs--;
            throw new ResetException("RESET");
        }
    }
}